### PR TITLE
feat: Choose a user during the Devtool login process

### LINKED ISSUE #211: [Task] Create a user selection page in the devtool OAuth process
## What should be done 
When using external clients we can't choose other user than the default one.

We should add an interface to select a user (list all the created user and a `+` to create a new one).

## Technical recommandation
Change the OAuth workflow to display page to select the user.

### PR DESCRIPTION
**Pull request by taorepoara**
## About this PR

Closes #211 

## Technical highlight/advice
The fake Hydra endpoint now manage views to rendre heex files.

The user id given through the socket overrides the token one to make it easy for Lenra client.

## How to test my changes
Run an app with the devtool following the README instructions and you'll see the user choosing page.

## Checklist
- [x] I didn't over-scope my PR
- [x] My PR title matches the [commit convention](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] I did not include breaking changes
- [x] I made my own code-review before requesting one

### I included unit tests that cover my changes
- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

### I added/updated the documentation about my changes
- [ ] 📜 README.md
- [ ] 📕 docs/*.md
- [ ] 📓 docs.lenra.io
- [x] 🙅 no documentation needed

### CODE REVIEW

File: server/lib/dev_tool/fake_hydra/controllers/oauth2_controller.ex [hunk 1]
```diff
@@ -1,13 +1,38 @@
...
-    code = OAuth2Helper.generate_code(params["scope"])
-    encoded_params = URI.encode_query(%{"code" => code, "state" => params["state"]})
+  def auth(conn, %{"redirect_uri" => redirect_uri, "scope" => scope, "state" => state}) do
+    users =
+      Enum.map(Repo.all(User), fn user -> user.manual_id end)
+      |> Enum.sort()
+
+    if Enum.empty?(users) do
+      {:ok, user} = UserServices.upsert_fake_user(1)
+      ^users = [1]
```
Comment (jonas-martinez) [inline server/lib/dev_tool/fake_hydra/controllers/oauth2_controller.ex:16]: Why are you using the pin operator "^" ? It is used to check if the variable matches with the right part but in this case the "users" variable is empty as per the if statement line 14.

File: server/lib/dev_tool/fake_hydra/router.ex [hunk 2]
```diff
@@ -4,8 +4,19 @@ defmodule DevTool.FakeHydra.Router do
...
+    # plug :fetch_live_flash
+    plug(:put_root_layout, {DevTool.LayoutView, :root})
+    # plug :protect_from_forgery
+    plug(:put_secure_browser_headers)
+  end
+
   scope "/", DevTool.FakeHydra do
+    pipe_through([:browser])
     get("/oauth2/auth", OAuth2Controller, :auth)
+    get("/choose/user", OAuth2Controller, :user)
```
Comment (jonas-martinez) [inline server/lib/dev_tool/fake_hydra/router.ex:19]: You could rename this route to match the path to the template file which is "oauth2/choose-user"

```suggestion
    get("/oauth2/choose-user", OAuth2Controller, :user)
```

File: server/lib/dev_tool/templates/fake_hydra/o_auth2/choose-user.html.heex [hunk 3]
```diff
@@ -0,0 +1,15 @@
+<% base_url = DevTool.FakeHydra.Router.Helpers.o_auth2_path(@conn, :user) %>
```
Comment (jonas-martinez) [inline server/lib/dev_tool/templates/fake_hydra/o_auth2/choose-user.html.heex:1]: You should rename the folder "o_auth2" to "oauth2" to keep the same naming across the project.

File: server/lib/dev_tool/templates/fake_hydra/o_auth2/choose-user.html.heex [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+<main class="arrows">
+    <nav>
+        <%= for user_id <- @users do %>
+            <a href={DevTool.FakeHydra.Router.Helpers.o_auth2_path(@conn, :user, user: user_id, redirect_uri: @redirect_uri, scope: @scope, state: @state)}>User <%= user_id %></a>
```
Comment (jonas-martinez) [inline server/lib/dev_tool/templates/fake_hydra/o_auth2/choose-user.html.heex:5]: After renaming the folder.

```suggestion
            <a href={DevTool.FakeHydra.Router.Helpers.oauth2_path(@conn, :user, user: user_id, redirect_uri: @redirect_uri, scope: @scope, state: @state)}>User <%= user_id %></a>
```

**Review by jonas-martinez — CHANGES_REQUESTED** [review]
Here is my first pass on this PR. 

I'm going to test the code to see if some of my comments are actually useful.

**Comment by jonas-martinez** [comment]
@taorepoara Where is the README explaining how to test this PR ? I don't understand how to change the user in the app.

File: server/lib/dev_tool/fake_hydra/controllers/oauth2_controller.ex (hunk 1 above)
Comment (taorepoara) [reply server/lib/dev_tool/fake_hydra/controllers/oauth2_controller.ex:16]: It's to override the value of the `users` var in the parent block (linter recommandation). I probably not understood it well...

File: server/lib/dev_tool/fake_hydra/router.ex (hunk 2 above)
Comment (taorepoara) [reply server/lib/dev_tool/fake_hydra/router.ex:19]: I did want to use oauth2 prefix in the route in order to do not have conflict with hydra routes.

**Comment by taorepoara** [comment]
> @taorepoara Where is the README explaining how to test this PR ? I don't understand how to change the user in the app.

To change the user, you need to have a disconnect button in the app.

I've added one in the js client lib example.

File: server/lib/dev_tool/fake_hydra/controllers/oauth2_controller.ex (hunk 1 above)
Comment (taorepoara) [reply server/lib/dev_tool/fake_hydra/controllers/oauth2_controller.ex:16]: image

**Review by jonas-martinez — APPROVED** [review]

**Pull request merged**